### PR TITLE
Adds synchronous span APIs for datagram sockets.

### PR DESCRIPTION
**Pull request by PJB3005**
Hope I did this right.

Part of #33418/#938

### CODE REVIEW

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs
Comment (gfoidl) [inline src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:1384]: Nit:
```suggestion
        /// Sends data to the specified endpoint.
```

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs
Comment (gfoidl) [inline src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:1398]:
```suggestion
        /// Sends data to a specific endpoint using the specified <see cref="SocketFlags"/>.
```

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Windows.cs
Comment (gfoidl) [inline src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Windows.cs:304]:
```suggestion
            SendTo(handle, buffer.AsSpan(offset, size), socketFlags, peerAddress, peerAddressSize, out bytesTransferred);
```

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Windows.cs
Comment (gfoidl) [inline src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Windows.cs:309]:
```suggestion
            fixed (byte* bufferPtr = buffer)
```
In the case `buffer` is empty this is guaranteed to return the null-ptr (to be precise: a Null-ref is returned which is then the null-ptr).

`MemoryMarshal.GetReference` doesn't provide that guarantee. So one should check for `IsEmpty` before. As it goes here to the kernel-call I'd go with the safe version.

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Windows.cs
Comment (gfoidl) [inline src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Windows.cs:516]:
```suggestion
            ReceiveFrom(handle, buffer.AsSpan(offset, size), SocketFlags.None, socketAddress, ref addressLength, out bytesTransferred);
```

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Windows.cs
Comment (gfoidl) [inline src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Windows.cs:522]:
```suggestion
            fixed (byte* bufferPtr = buffer)
```
See above.

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Windows.cs
Comment (PJB3005) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Windows.cs:309]: The span versions of `Send` and `Receive` [do the same thing](https://github.com/dotnet/runtime/blob/2223babdd49118787c675e04aff711f936a10b26/src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Windows.cs#L278) which I copied. Should I change those too?

Comment (gfoidl) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Windows.cs:309]: Had another thought about this. As you pass `buffer.Length` to the Win-API your code is OK, and please ignore my comment here.

FYI https://github.com/dotnet/runtime/issues/25138#issuecomment-375163012 handles that case is general.

**Comment by geoffkizer** [comment]
Thanks, this looks good to me.

@antonfirsov should review and approve. (I think he's out today but should be back.)

To be clear -- the ReceiveMessageFrom API is already done, right? I see it listed in #33418, but I think that's incorrect, right?

**Comment by PJB3005** [comment]
#46285 did `ReceiveMessageFrom`.

@antonfirsov is the way I did the tests correct? I didn't test manually (since I don't know how to compile against a local runtime like this).

Also the failing test seems to be a timeout in some HTTP API which I don't think I could have broken with these changes?

**Comment by geoffkizer** [comment]
> Also the failing test seems to be a timeout in some HTTP API which I don't think I could have broken with these changes?

Yeah, don't worry about the failing test. Test results look fine.

**Review by antonfirsov — COMMENTED** [review]
Product code looks good, although I have some questions to the team regarding `NetEventSource`.

Can we add 0-byte send tests similarly to #51473?

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs
Comment (geoffkizer) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:1416]: NetEventSource is whatever we want for internal logging purposes. It's not intended to be public, ever. To the extent that it duplicates our public telemetry, we should remove it. But there's a lot of internal logging code that has proven useful in the past, and we don't want to just blindly rip this out.

Comment (antonfirsov) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:1416]: OK, from what you said it feels like it's not worth opening an issue for it, but we should keep it in mind. A dedupe refactor across the span/array overloads should fix this anyways.

Comment (geoffkizer) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:1416]: Yep.

> A dedupe refactor across the span/array overloads should fix this anyways.

This is blocked on #46600 right?

**Comment by geoffkizer** [comment]
What else needs to happen here before we merge? @antonfirsov?

**Comment by antonfirsov** [comment]
In my https://github.com/dotnet/runtime/pull/51956#pullrequestreview-648356203 I recommended to add tests for the 0-byte send tests similarly to #51473. Not essential, but would be nice to have.

**Comment by PJB3005** [comment]
Ah, my bad, didn't realize that was something I should do.

**Comment by PJB3005** [comment]
I added 0-byte send tests. Do these look good @antonfirsov?

**Review by antonfirsov — APPROVED** [review]
LGTM. Test failures are unrelated.

Thanks!

**Pull request merged**
